pkg/api/resource/quantity.go

```diff
@@ -0,0 +1,406 @@
+	"math/big"
+	"regexp"
+	"strings"
+
+	"speter.net/go/exp/math/dec/inf"
+)
+
+// Quantity is a fixed-point representation of a number.
+// It provides convenient marshaling/unmarshaling in JSON and YAML,
+// in addition to String() and Int64() accessors.
+//
+// The serialization format is:
+//
+// <serialized>      ::= <sign><numeric> | <numeric>
+// <numeric>         ::= <digits><suffix> | <digits>.<digits><suffix>
+//   (Note that <suffix> may be ""!)
+// <sign>            ::= "+" | "-"
+// <digits>          ::= <digit> | <digit><digits>
+// <signedDigits>    ::= <digits> | <sign><digits>
+// <digit>           ::= 0 | 1 | ... | 9
+// <suffix>          ::= <binarySuffix> | <decimalExponent> | <decimalSuffix>
+// <binarySuffix>    ::= i | Ki | Mi | Gi | Ti | Pi | Ei
+// <decimalSuffix>   ::= m | "" | k | M | G | T | P | E
+// <decimalExponent> ::= "e" <signedDigits> | "E" <signedDigits>
+//   (Note that 1024 = 1Ki but 1000 = 1k; I didn't choose the capitalization.)
+//
+// No matter which of the three exponent forms is used, no quantity may represent
+// a number greater than 2^63-1 in magnitude, nor may it have more than 3 digits
+// of precision. Numbers larger or more precise will be capped or rounded.
+// (E.g.: 0.1m will rounded up to 1m.)
+// This may be extended in the future if we require larger or smaller quantities.
+//
+// When a Quantity is parsed from a string, it will remember the type of suffix
+// it had, and will use the same type again when it is serialized.
+// One exception: numbers with a Binary SI suffix less than one will be changed
+// to Decimal SI suffix. E.g., .5i becomes 500m. [NOT 512m!]
+//
+// Before serializing, Quantity will be put in "canonical form".
+// This means that Exponent/suffix will be adjusted up or down (with a
+// corresponding increase or decrease in Mantissa) such that:
+//   a. No precision is lost
+//   b. No fractional digits will be emitted
+//   c. The exponent (or suffix) is as large as possible.
+// The sign will be omitted unless the number is negative.
+//
+// Examples:
+//   1.5 will be serialized as "1500m"
+//   1.5Gi will be serialized as "1576Mi"
+//
+// Note that the quantity will NEVER be internally represented by a
+// floating point number. That is the whole point of this exercise.
+//
+// Non-canonical values will still parse as long as they are well formed,
+// but will be re-emitted in their canonical form. (So always use canonical
+// form, or don't diff.)
+//
+// This format is intended to make it difficult to use these numbers without
+// writing some sort of special handling code in the hopes that that will
+// cause implementors to also use a fixed point implementation.
+type Quantity struct {
+	// Amount is public, so you can manipulate it if the accessor
+	// functions are not sufficient.
+	Amount *inf.Dec
+
+	// Change Format at will. See the comment for Canonicalize for
+	// more details.
+	Format
+}
+
+// Format lists the three possible formattings of a quantity.
```

Comment:
I don't think changing the names is sufficient; I think it would be helpful to give some concrete examples of which constructor you should use for each of the above quantities, and what Format would that end up getting encoded as. As-is this class is sufficiently abstracted from actual cluster management quantities that I think people may find it confusing. Examples can bridge the gap.